tmol/io/pose_stack_construction.py

```diff
@@ -0,0 +1,195 @@
+        termini residues, they will not be built with their termini-variant
+        types. The purpose is to allow the user to include a subset of the
+        residues in a protein where a series of "gap" residues can be omitted
+        between i and i+1 without those two residues being treated as if they
+        are chemically bonded. This will keep the Ramachandran term from scoring
+        nonsense dihdral angles and will keep the cart-bonded term from scoring
+        nonsense bond lengths and angles.
+
+    return_chain_ind: return the chain-index tensor that has been "left-justified"
+        from the chain
+    """
+
+    from tmol.io.details.canonical_packed_block_types import (
+        default_canonical_packed_block_types,
+    )
+    from tmol.io.details.left_justify_canonical_form import left_justify_canonical_form
+    from tmol.io.details.disulfide_search import find_disulfides
+    from tmol.io.details.his_taut_resolution import resolve_his_tautomerization
+    from tmol.io.details.select_from_canonical import (
+        assign_block_types,
+        take_block_type_atoms_from_canonical,
+    )
+    from tmol.io.details.build_missing_leaf_atoms import build_missing_leaf_atoms
+
+    assert chain_id.device == res_types.device
+    assert chain_id.device == coords.device
+
+    # step 1: retrieve the global packed_block_types object with the 66
+    #         canonical residue types
+    # step 2: remove any "virtual residues," marked with a res-type ind of -1
+    #         by shifting all of the residues in each Pose "to the left"
+    # step 3: resolve disulfides
+    # step 4: resolve his tautomer
+    # step 5: resolve termini variants, assign block-types to each input
+    #         residue, and populate the inter-block connectivity tensors
+    # step 6: select the atoms from the canonically-ordered input tensors
+    #         (the coords and atom_is_present tensors) that belong to the
+    #         now-assigned block types, discarding/ignoring
+    #         any others that may have been provided
+    # step 7: if any atoms missing, build them
+    # step 8: construct PoseStack object
+
+    if atom_is_present is None:
+        atom_is_present = torch.all(torch.logical_not(torch.isnan(coords)), dim=3)
+
+    # 1
+    # this will return the same object each time to minimize the number
+    # of calls to the setup_packed_block_types annotation functions
+    pbt, atom_type_resolver = default_canonical_packed_block_types(chain_id.device)
+
+    # 2
+    # "left justify" the input canonical-form residues: residues that are
+    (
+        chain_id,
+        res_types,
+        coords,
+        atom_is_present,
+        disulfides,
+        res_not_connected,
+    ) = left_justify_canonical_form(
+        chain_id, res_types, coords, atom_is_present, disulfides, res_not_connected
+    )
+
+    # 3
+    found_disulfides, res_type_variants = find_disulfides(res_types, coords, disulfides)
```

Comment:
ok, if i've read the rest of this correctly, this function takes input disulfides and then runs detection looking for additional ones.  There should be a way to specify that detection shouldn't add new dslfs as well (e.g., even though SGs are close, NO dslfs should be "found")